calico_kubernetes/calico_kubernetes.py

```diff
@@ -97,6 +147,223 @@ def _configure_interface(self):
+        netns.remove_veth(endpoint.name)
+
+        # Remove the container from the datastore.
+        self._datastore_client.remove_workload(hostname, ORCHESTRATOR_ID, workload_id)
+
+        print("Removed Calico interface from %s" % container_name)
+
+    def _validate_container_add(self, info):
+        container_id = info["Id"]
+
+        # Check if the container already exists
+        try:
+            _ = self._datastore_client.get_endpoint(
+                hostname=hostname,
+                orchestrator_id=ORCHESTRATOR_ID,
+                workload_id=container_id
+            )
+        except KeyError:
+            # Calico doesn't know about this container.  Continue.
+            pass
+        else:
+            # Calico already set up networking for this container.  Since we got
+            # called with an IP address, we shouldn't just silently exit, since
+            # that would confuse the user: the container would not be reachable on
+            # that IP address.
+            print("This container has already been configured with Calico Networking.")
+            sys.exit(1)
+
+        # Check the container is actually running.
+        if not info["State"]["Running"]:
+            print("The container is not currently running.")
+            sys.exit(1)
+
+        # We can't set up Calico if the container shares the host namespace.
+        if info["HostConfig"]["NetworkMode"] == "host":
+            print("Can't add the container to Calico because it is running NetworkMode = host.")
+            sys.exit(1)
+
+    def _assign_address_in_pool_IPAM(self, ip):
+        # Get pool
+        pools = self._datastore_client.get_ip_pools(ip.version)
+        pool = None
+        for candidate_pool in pools:
+            if ip in candidate_pool:
+                pool = candidate_pool
+                break
+        if pool is None:
+            print("%s is not in any configured pools" % ip)
+            sys.exit(1)
+
+        # Assign the IP
+        if not self._ipam_client.assign_address(pool, ip):
+            print("IP address is already assigned in pool %s " % pool)
+            sys.exit(1)
+
+    def _unassign_address_in_pool_IPAM(self, ip):
+        pools = self._datastore_client.get_ip_pools(ip.version)
+        for pool in pools:
+            if ip in pool:
+                # Ignore failure to unassign address, since we're not
+                # enforcing assignments strictly in datastore.py.
+                self._ipam_client.unassign_address(pool, ip)
+
+    def _endpoint_init(self, container_id, ip):
+         # The next hop IPs for this host are stored in etcd.
+        next_hops = self._datastore_client.get_default_next_hops(hostname)
+        try:
+            next_hop = next_hops[ip.version]
+        except KeyError:
+            print("This node is not configured for IPv%d." % ip.version)
+            sys.exit(1)
+
+        network = IPNetwork(IPAddress(ip))
+        ep = Endpoint(hostname=hostname,
```

Comment:
hostname and ORCHESTRATOR_ID are only available in this context because they are globals in the calico_kubernetes file.  If this gets moved to pycalico, we'll need to pass in the ORCHESTRATOR_ID and the hostname as arguments.